flake.nix

```diff
@@ -153,124 +153,55 @@
-            inherit
-              fileset
-              stdenv
-              versionSuffix
-              ;
-          };
-
-          nix-util-c = final.callPackage ./src/libutil-c/package.nix {
-            inherit
-              fileset
-              stdenv
-              versionSuffix
-              ;
-          };
+          # A new scope, so that we can use `callPackage` to inject our own interdependencies
+          # without "polluting" the top level "`pkgs`" attrset.
+          # This also has the benefit of providing us with a distinct set of packages
+          # we can iterate over.
+          nixComponents = lib.makeScope final.nixDependencies.newScope (import ./packaging/components.nix {
+            pkgs = final;
+            inherit stdenv versionSuffix;
+          });
 
-          nix-store = final.callPackage ./src/libstore/package.nix {
-            inherit
-              fileset
-              stdenv
-              officialRelease
-              versionSuffix
-              ;
-            libseccomp = final.libseccomp-nix;
+          # The dependencies are in their own scope, so that they don't have to be
+          # in Nixpkgs top level `pkgs` or `nixComponents`.
+          nixDependencies = lib.makeScope final.newScope (scope: {
+            inherit stdenv versionSuffix;
+            libseccomp = final.libseccomp_nix;
+            boehmgc = final.boehmgc_nix;
+            libgit2 = final.libgit2_nix;
             busybox-sandbox-shell = final.busybox-sandbox-shell or final.default-busybox-sandbox-shell;
-          };
-
-          nix-fetchers = final.callPackage ./src/libfetchers/package.nix {
-            inherit
-              fileset
-              stdenv
-              officialRelease
-              versionSuffix
-              ;
-          };
-
-          nix =
-            final.callPackage ./package.nix {
-              inherit
-                fileset
-                stdenv
-                officialRelease
-                versionSuffix
-                ;
-              boehmgc = final.boehmgc-nix;
-              libgit2 = final.libgit2-nix;
-              libseccomp = final.libseccomp-nix;
-              busybox-sandbox-shell = final.busybox-sandbox-shell or final.default-busybox-sandbox-shell;
-            };
+          });
 
-          nix-perl-bindings = final.callPackage ./src/perl/package.nix {
-            inherit
-              fileset
-              stdenv
-              versionSuffix
-              ;
-          };
+          nix = final.nixComponents.nix;
 
           nix-internal-api-docs = final.callPackage ./src/internal-api-docs/package.nix {
             inherit
-              fileset
               stdenv
               versionSuffix
               ;
           };
 
           nix-external-api-docs = final.callPackage ./src/external-api-docs/package.nix {
             inherit
-              fileset
               stdenv
               versionSuffix
               ;
```

Comment:
Yeah, we're not using `attrNames nixComponents` anyway, so it's actually fine.
Early on in the commit series it would have been suitable for that, but `makeScope` adds some stuff that'd need to be filtered out, which I think is the right trade-off.